knowledge_repo/app/index.py

```diff
@@ -1,17 +1,99 @@
 from __future__ import absolute_import
 import logging
-from .proxies import db_session, current_repo
-from .models import Post
+import datetime
+import threading
+from flask import has_app_context
+
+from .proxies import db_session, current_repo, current_app
+from .models import Post, IndexMetadata
 from .utils.emails import send_subscription_emails
 from .utils.search import get_keywords
 
 logging.basicConfig(level=logging.INFO)
 logger = logging.getLogger(__name__)
 
 
+def is_indexing():
+    return int(IndexMetadata.get('lock', 'index', '0'))
+
+
+def seconds_since_index():
+    last_update = IndexMetadata.get_last_update('lock', 'index')
+    if last_update is None:
+        return None
+    else:
+        return (datetime.datetime.utcnow() - last_update).total_seconds()
+
+
+def seconds_since_index_check():
+    last_update = IndexMetadata.get_last_update('check', 'index')
+    if last_update is None:
+        return None
+    else:
+        return (datetime.datetime.utcnow() - last_update).total_seconds()
+
+
+def human_readable_time_since_index():
+    if is_indexing():
+        return 'Currently indexing'
+    seconds = seconds_since_index()
+    if seconds is None:
+        return "Never"
+    elif seconds < 60:
+        return "{:d} seconds ago".format(int(round(seconds)))
+    else:
+        return "{:d} minutes ago".format(int(round(seconds / 60)))
+
+
+def update_index_required():
+    if not current_app.config.get('REPOSITORY_INDEXING_ENABLED', True):
+        return False
+
+    seconds = seconds_since_index()
+    seconds_check = seconds_since_index_check()
+    if is_indexing() or (seconds is not None) and (seconds < 5 * 60) and (seconds_check < 5 * 60):
+        return False
+    try:
```

Comment:
I think this is actually kosher Python code. It's the neatest way to guarantee that code block always runs even if there are exceptions (which are not caught).